home/tests/test_display_question_feature.py

```diff
@@ -107,10 +91,97 @@ def test_response_context(self, response):
+
+    class TestLikesMechanism:
+        @pytest.mark.django_db
+        def test_logged_user_like(self, client):
+            """
+            in this test, a logged user press like on (answer A) which he disliked before
+            """
+            user = Profile.objects.all().get(user_id=1).user
+            client.force_login(user)
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "like 9"})
+            response = client.get(response.url)
+            answerA = '(<Answer: Answer A>, \'bi bi-hand-thumbs-up-fill\', \'bi bi-hand-thumbs-down'
+            assert answerA in str(response.context['answers_structure'])
+
+        @pytest.mark.django_db
+        def test_logged_user_dislike(self, client):
+            """
+            in this test, a logged user press dislike on (answer B) which he liked before
+            """
+            user = Profile.objects.all().get(user_id=1).user
+            client.force_login(user)
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "dislike 10"})
+            response = client.get(response.url)
+            answerB = '(<Answer: Answer B>, \'bi bi-hand-thumbs-up\', \'bi bi-hand-thumbs-down-fill'
+            assert answerB in str(response.context['answers_structure'])
+
+        @pytest.mark.django_db
+        def test_logged_user_unlike(self, client):
+            """
+            in this test, a logged user press like on (answer B) which he liked before
+            and make it un-liked
+            """
+            user = Profile.objects.all().get(user_id=1).user
+            client.force_login(user)
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "like 10"})
+            response = client.get(response.url)
+            answerB = '(<Answer: Answer B>, \'bi bi-hand-thumbs-up\', \'bi bi-hand-thumbs-down'
+            assert answerB in str(response.context['answers_structure'])
+
+        @pytest.mark.django_db
+        def test_logged_user_undislike(self, client):
+            """
+            in this test, a logged user press dislike on (answer A) which he disliked before
+            and make it un-disliked
+            """
+            user = Profile.objects.all().get(user_id=1).user
+            client.force_login(user)
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "dislike 9"})
+            response = client.get(response.url)
+            answerA = '(<Answer: Answer A>, \'bi bi-hand-thumbs-up\', \'bi bi-hand-thumbs-down'
+            assert answerA in str(response.context['answers_structure'])
+
+        @pytest.mark.django_db
+        def test_invalid_thumbs_call_from_unsigned_user(self, client):
+            """
+            Unsigned user should not be able to like an answer
+            """
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "like 9"})
+            assert response.status_code == 404
```

Comment:
Maybe you should add redirection to a log-in page in this situation instead of to the error page.
What do you think?

Reply:
I think I don't want to do it because:
The thumbs view should be reached only from a legit like/dislike button press of a signed user.
By redirecting to the login page when entering a URL that shouldn't be entered manually is a way of saying - what you did is a legit and acceptable operation, and let me help you sign in so you could complete this operation. but I don't want to allow it. Moreover, this test is testing a case that will never happen because the thumbs view needs to get an answer ID from a post request, so this is another reason it should fail.
Unsigned users have a link for login under the disabled like buttons, as you can see in the picture in the issue linked to this PR.

Reply:
It would probably be better to respond with something like 401 (unauthorized) or 403 (forbidden) then 404 (not found):

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401